openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/keystore.py

```diff
@@ -3,100 +3,178 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 
-"""The keystore module contains an interface to store and retrieve certificates
-and keys used for signing.
+"""
+This module contains an interface for generating code signing credentials.
+
+The interface sends HTTP requests to a REST API server.
 
-The KeyStore class wraps a backend class and delegates all calls to it.
+The REST API forwards the calls to a "backend" secrets engine,
+which will do the real work of generating and storing the keys.
 
-The backend is the interface to the secrets engine you have chosen to use.
-The backend class will typically be a wrapper around some REST api calls.
-To use the KeyStore, instantiate a backend and pass it to the KeyStore
-constructor.
+The secrets engine interface is defined in the signing.pki.backends
+module.
 """
 
 import json
 import logging
+import multiprocessing
+import pathlib
+import subprocess
+import time
+
+import requests
 
+from signing.pki import endpoints
 
 logger = logging.getLogger("mbl-signing.pki")
 
 
+class KeyStoreError(Exception):
+    """API call failed."""
+
+
 class KeyStoreResponse:
     """Convert HTTPResponse json objects from the backend to an object."""
+
     def __init__(self, response):
-        self.raw = response.text
-        self._data = None
-        try:
-            self._data = json.loads(self.raw)
-        except json.decoder.JSONDecodeError:
-            # The response content is empty, just return an empty instance.
-            return
+        """
+        Initialise the response.
+
+        Take a HTTPResponse and iterate over the json object.
+        Set attributes on this class from the json object.
+
+        :params response HTTPResponse: A raw HTTPResponse.
+        """
+        if isinstance(response, dict):
+            self._data = response
+        else:
+            try:
+                self._data = json.loads(response.text)
+            except json.decoder.JSONDecodeError:
+                # The response content is empty, just return an empty instance.
+                return
         for k, v in self._data.items():
             setattr(self, k, v)
+        if hasattr(self, "status") and self.status != 200:
+            msg = "Returned status: {}\n\nError message from server: {}\n"
+            raise KeyStoreError(msg.format(self.status, self.detail))
+
+
+def run_server_proc():
+    """Run the REST api server as a subprocess."""
+    proc = multiprocessing.Process(target=endpoints.main)
+    proc.start()
+    time.sleep(1)
```

Comment:
Is this a pause to wait for the process to begin?  Or a work-around for an issue?

Reply:
It's a pause to wait for the process to begin and the server to start up. Not the nicest solution admittedly, I will add a task to change it in the next version.